File: leetcode/0239_sliding_window_maximum.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <deque>
#include <vector>

using namespace std;

class Solution {
 public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;

        deque<int> deque;
        for (int i = 0; i < nums.size(); i++) {
            while (!deque.empty() && deque.front() <= i - k) {
                deque.pop_front();
            }

            while (!deque.empty() && nums[i] >= nums[deque.back()]) {
                deque.pop_back();
            }

            deque.push_back(i);

            if (i >= k - 1) {
                result.push_back(nums[deque.front()]);
            }
        }

        return result;
    }
};

int main() {
  auto solution = Solution();

  vector<int> nums = {1, 3, -1, -3, 5, 3, 6, 7};

  vector<int> expected = {3, 3, 5, 5, 6, 7};
  vector<int> result = solution.maxSlidingWindow(nums, 3);

  assert(expected == result);
}

File: leetcode/0239_sliding_window_maximum.py
# -*- coding: utf-8 -*-

from collections import deque


class Solution:
    def maxSlidingWindow(self, nums, k):
        result = []
        seen = deque()

        for i, num in enumerate(nums):
            while seen and num >= nums[seen[-1]]:
                seen.pop()
            seen.append(i)
            if i >= k and seen and seen[0] == i - k:
                seen.popleft()
            if i >= k - 1:
                result.append(nums[seen[0]])

        return result


if __name__ == '__main__':
    solution = Solution()

    assert [3, 3, 5, 5, 6, 7] == solution.maxSlidingWindow([
        1, 3, -1, -3, 5, 3, 6, 7], 3)
